app/services/supplementary-billing/reissue-invoice.service.js

```diff
@@ -0,0 +1,267 @@
+  return chargingModuleLicence.transactions.find((transaction) => {
+    return transaction.rebilledTransactionId === id
+  })
+}
+
+function _retrieveChargingModuleLicence (chargingModuleInvoice, licenceRef) {
+  return chargingModuleInvoice.licences.find((licence) => {
+    return licence.licenceNumber === licenceRef
+  })
+}
+
+/**
+ * If a billing invoice exists for this combination of source invoice and CM reissue invoice then return it; otherwise,
+ * generate it, store it and then return it.
+ */
+function _retrieveOrGenerateBillingInvoice (dataToReturn, sourceInvoice, reissueBillingBatch, chargingModuleReissueInvoice) {
+  // Because we have nested iteration of source invoice and Charging Module reissue invoice, we need to ensure we have
+  // a billing invoice for every combination of these, hence we search by both of their ids
+  const existingBillingInvoice = dataToReturn.billingInvoices.find((invoice) => {
+    return invoice.invoiceAccountId === sourceInvoice.invoiceAccountId &&
+      invoice.externalId === chargingModuleReissueInvoice.id
+  })
+
+  if (existingBillingInvoice) {
+    return existingBillingInvoice
+  }
+
+  const translatedChargingModuleInvoice = _mapChargingModuleInvoice(chargingModuleReissueInvoice)
+  const generatedBillingInvoice = GenerateBillingInvoiceService.go(
+    sourceInvoice,
+    reissueBillingBatch.billingBatchId,
+    sourceInvoice.financialYearEnding
+  )
+
+  // Construct our new billing invoice by taking the generated billing invoice and combining it with the mapped fields
+  // from the CM invoice, then updating its `originalBillingInvoiceId` field
+  const newBillingInvoice = {
+    ...generatedBillingInvoice,
+    ...translatedChargingModuleInvoice,
+    originalBillingInvoiceId: sourceInvoice.billingInvoiceId
+  }
+
+  dataToReturn.billingInvoices.push(newBillingInvoice)
+
+  return newBillingInvoice
+}
+
+/**
+ * If a billing invoice licence exists for this invoice account id then return it; otherwise, generate it, store it and
+ * then return it.
+ */
+function _retrieveOrGenerateBillingInvoiceLicence (dataToReturn, sourceInvoice, billingInvoiceId, sourceInvoiceLicence) {
+  const existingBillingInvoiceLicence = dataToReturn.billingInvoiceLicences.find((invoice) => {
+    return invoice.invoiceAccountId === sourceInvoice.invoiceAccountId
+  })
+
+  if (existingBillingInvoiceLicence) {
+    return existingBillingInvoiceLicence
+  }
+
+  const newBillingInvoiceLicence = GenerateBillingInvoiceLicenceService.go(billingInvoiceId, sourceInvoiceLicence)
+
+  dataToReturn.billingInvoiceLicences.push(newBillingInvoiceLicence)
+
+  return newBillingInvoiceLicence
+}
+
+async function _sendReissueRequest (billingBatchExternalId, invoiceExternalId) {
+  try {
+    const result = await ChargingModuleReissueInvoiceService.go(billingBatchExternalId, invoiceExternalId)
+    return result.response.invoices
+  } catch (error) {
+    throw new Error(`Charging Module reissue request failed: ${error.message}`)
```

Comment:
I'm not sure we would ever actually hit these error blocks. The ChargingModule*Services all depend on `ChargingModuleRequestLib` which in turn depends on `RequestLib`.

They don't add much logic between the call and what `Requestlib` does and `RequestLib` ensures all errors are caught, logged, and then the result object that's returned is populated with the error.

We should properly be checking `result.succeeded` and then deciding whether an error needs to be logged and thrown.